Getting the following error when navigating to the "/addbook" page on this branch:

```
Whitelabel Error Page
This application has no explicit mapping for /error, so you are seeing this as a fallback.

Thu Mar 28 23:50:05 EDT 2019
There was an unexpected error (type=Internal Server Error, status=500).
An error happened during template parsing (template: "class path resource [templates/create-book.html]")
org.thymeleaf.exceptions.TemplateInputException: An error happened during template parsing (template: "class path resource [templates/create-book.html]")
	at org.thymeleaf.templateparser.markup.AbstractMarkupTemplateParser.parse(AbstractMarkupTemplateParser.java:241)
	at org.thymeleaf.templateparser.markup.AbstractMarkupTemplateParser.parseStandalone(AbstractMarkupTemplateParser.java:100)
	at org.thymeleaf.engine.TemplateManager.parseAndProcess(TemplateManager.java:666)
	at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1098)
	at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1072)
	at org.thymeleaf.spring5.view.ThymeleafView.renderFragment(ThymeleafView.java:362)
	at org.thymeleaf.spring5.view.ThymeleafView.render(ThymeleafView.java:189)
	at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1370)
	at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1116)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1055)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:897)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:200)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:834)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1415)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.attoparser.ParseException: Error during execution of processor 'org.thymeleaf.spring5.processor.SpringInputGeneralFieldTagProcessor' (template: "create-book" - line 11, col 60)
	at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:393)
	at org.attoparser.MarkupParser.parse(MarkupParser.java:257)
	at org.thymeleaf.templateparser.markup.AbstractMarkupTemplateParser.parse(AbstractMarkupTemplateParser.java:230)
	... 52 more
Caused by: org.thymeleaf.exceptions.TemplateProcessingException: Error during execution of processor 'org.thymeleaf.spring5.processor.SpringInputGeneralFieldTagProcessor' (template: "create-book" - line 11, col 60)
	at org.thymeleaf.processor.element.AbstractAttributeTagProcessor.doProcess(AbstractAttributeTagProcessor.java:117)
	at org.thymeleaf.processor.element.AbstractElementTagProcessor.process(AbstractElementTagProcessor.java:95)
	at org.thymeleaf.util.ProcessorConfigurationUtils$ElementTagProcessorWrapper.process(ProcessorConfigurationUtils.java:633)
	at org.thymeleaf.engine.ProcessorTemplateHandler.handleStandaloneElement(ProcessorTemplateHandler.java:918)
	at org.thymeleaf.engine.StandaloneElementTag.beHandled(StandaloneElementTag.java:228)
	at org.thymeleaf.engine.Model.process(Model.java:282)
	at org.thymeleaf.engine.ProcessorTemplateHandler.handleOpenElement(ProcessorTemplateHandler.java:1587)
	at org.thymeleaf.engine.OpenElementTag.beHandled(OpenElementTag.java:205)
	at org.thymeleaf.engine.Model.process(Model.java:282)
	at org.thymeleaf.engine.ProcessorTemplateHandler.handleOpenElement(ProcessorTemplateHandler.java:1587)
	at org.thymeleaf.engine.OpenElementTag.beHandled(OpenElementTag.java:205)
	at org.thymeleaf.engine.Model.process(Model.java:282)
	at org.thymeleaf.engine.Model.process(Model.java:290)
	at org.thymeleaf.engine.GatheringModelProcessable.process(GatheringModelProcessable.java:78)
	at org.thymeleaf.engine.ProcessorTemplateHandler.handleCloseElement(ProcessorTemplateHandler.java:1640)
	at org.thymeleaf.engine.TemplateHandlerAdapterMarkupHandler.handleCloseElementEnd(TemplateHandlerAdapterMarkupHandler.java:388)
	at org.thymeleaf.templateparser.markup.InlinedOutputExpressionMarkupHandler$InlineMarkupAdapterPreProcessorHandler.handleCloseElementEnd(InlinedOutputExpressionMarkupHandler.java:322)
	at org.thymeleaf.standard.inline.OutputExpressionInlinePreProcessorHandler.handleCloseElementEnd(OutputExpressionInlinePreProcessorHandler.java:220)
	at org.thymeleaf.templateparser.markup.InlinedOutputExpressionMarkupHandler.handleCloseElementEnd(InlinedOutputExpressionMarkupHandler.java:164)
	at org.attoparser.HtmlElement.handleCloseElementEnd(HtmlElement.java:169)
	at org.attoparser.HtmlMarkupHandler.handleCloseElementEnd(HtmlMarkupHandler.java:412)
	at org.attoparser.MarkupEventProcessorHandler.handleCloseElementEnd(MarkupEventProcessorHandler.java:473)
	at org.attoparser.ParsingElementMarkupUtil.parseCloseElement(ParsingElementMarkupUtil.java:201)
	at org.attoparser.MarkupParser.parseBuffer(MarkupParser.java:725)
	at org.attoparser.MarkupParser.parseDocument(MarkupParser.java:301)
	... 54 more
Caused by: java.lang.IllegalStateException: Neither BindingResult nor plain target object for bean name 'book' available as request attribute
	at org.springframework.web.servlet.support.BindStatus.<init>(BindStatus.java:153)
	at org.springframework.web.servlet.support.RequestContext.getBindStatus(RequestContext.java:903)
	at org.thymeleaf.spring5.context.webmvc.SpringWebMvcThymeleafRequestContext.getBindStatus(SpringWebMvcThymeleafRequestContext.java:227)
	at org.thymeleaf.spring5.util.FieldUtils.getBindStatusFromParsedExpression(FieldUtils.java:306)
	at org.thymeleaf.spring5.util.FieldUtils.getBindStatus(FieldUtils.java:253)
	at org.thymeleaf.spring5.util.FieldUtils.getBindStatus(FieldUtils.java:227)
	at org.thymeleaf.spring5.processor.AbstractSpringFieldTagProcessor.doProcess(AbstractSpringFieldTagProcessor.java:174)
	at org.thymeleaf.processor.element.AbstractAttributeTagProcessor.doProcess(AbstractAttributeTagProcessor.java:74)
	... 78 more
```

--- a/src/main/java/amazin/BuildSearchIndex.java
+++ b/src/main/java/amazin/BuildSearchIndex.java
@@ -1,0 +1,40 @@
+package amazin;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.transaction.Transactional;
+
+import org.hibernate.search.jpa.FullTextEntityManager;
+import org.hibernate.search.jpa.Search;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.ApplicationListener;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class is used to build the Lucene index at application startup. This is
+ * needed when the database has data before the web application is started.
+ * 
+ */
+@Component
+@Transactional
+public class BuildSearchIndex implements ApplicationListener<ApplicationReadyEvent> {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    /**
+     * Create an initial Lucene index for the data already present in the database.
+     * This method is called when Spring's startup.
+     */
+    @Override
+    public void onApplicationEvent(final ApplicationReadyEvent event) {
+        try {
+            FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(entityManager);
+            fullTextEntityManager.createIndexer().startAndWait();
+        } catch (InterruptedException e) {
+            System.out.println("An error occurred trying to build the serach index: " + e.toString());
+        }
+        return;
+    }
+
+}

--- a/src/main/java/amazin/controller/ApplicationController.java
+++ b/src/main/java/amazin/controller/ApplicationController.java
@@ -15,27 +15,27 @@ public class ApplicationController {
 
     @GetMapping({ "/" })
     public String index(Model model) {
-        model.addAttribute("books", bookRepository.findAll());
+        model.addAttribute(BookController.MODEL_ATTRIBUTE_BOOK, bookRepository.findAll());
         return "index";
     }
 
-    @GetMapping({"/login"})
-    public String login(){
+    @GetMapping({ "/login" })
+    public String login() {
         return "login";
     }
 
-    @GetMapping({"/register"})
-         public String register(){
+    @GetMapping({ "/register" })
+    public String register() {
         return "registration";
     }
 
-    @GetMapping({"/admin/register"})
-    public String registerAdmin(){
+    @GetMapping({ "/admin/register" })
+    public String registerAdmin() {
         return "registration-admin";
     }
 
-    @GetMapping({"/forgot-password"})
-    public String forgotPassword(){
+    @GetMapping({ "/forgot-password" })
+    public String forgotPassword() {
         return "forgot-password";
     }
 

--- a/src/main/java/amazin/controller/BookController.java
+++ b/src/main/java/amazin/controller/BookController.java
@@ -110,9 +110,8 @@ public class BookController {
         try {
             searchResults = searchservice.fuzzySearch(text, fieldNames);
         } catch (Exception ex) {
+            /** TODO: handle this exception properly and display message to user */
             ex.printStackTrace();
-            // for now do nothing. Later we will throw an exception and display a message to
-            // the user
         }
         if (searchResults != null && searchResults.isEmpty())
             model.addAttribute(MODEL_ATTRIBUTE_BOOK, null);

--- a/src/main/java/amazin/controller/BookController.java
+++ b/src/main/java/amazin/controller/BookController.java
@@ -22,11 +22,11 @@ import amazin.service.BookService;
 import amazin.service.HibernateSearchService;
 
 @Controller
-@SessionAttributes("book")
+@SessionAttributes("books")
 public class BookController {
     public static final String VIEW_CREATE_BOOK = "create-book";
     public static final String VIEW_UPDATE_BOOK = "update-book";
-    public static final String MODEL_ATTRIBUTE_BOOK = "book";
+    public static final String MODEL_ATTRIBUTE_BOOK = "books";
     public static final String PARAMETER_BOOK_ID = "id";
     public static final String REQUEST_MAPPING_BOOK = "/";
 
@@ -97,48 +97,28 @@ public class BookController {
         return createRedirectViewPath(REQUEST_MAPPING_BOOK);
     }
 
-    @GetMapping("/fuzzySearchName")
-    public String fuzzySearchByBookName(@RequestParam(value = "keywords", required = false) String text, Model model) {
+    @GetMapping("/search")
+    public String searchByBookName(@RequestParam(value = "keywords", required = true) String text, Model model) {
+
+        if (text.isEmpty())
+            return createRedirectViewPath(REQUEST_MAPPING_BOOK);
+
+        String[] fieldNames = { Book.BOOK_NAME_FIELD, Book.BOOK_DESCRIPTION_FIELD, Book.BOOK_ISBN_FIELD,
+                Book.BOOK_AUTHOR_FIELD, Book.BOOK_PUBLISHER_FIELD };
+
         List<Book> searchResults = null;
         try {
-            searchResults = searchservice.fuzzySearch(text, "name");
-
+            searchResults = searchservice.fuzzySearch(text, fieldNames);
         } catch (Exception ex) {
+            ex.printStackTrace();
             // for now do nothing. Later we will throw an exception and display a message to
             // the user
         }
-        model.addAttribute(MODEL_ATTRIBUTE_BOOK, searchResults);
-        return createRedirectViewPath(REQUEST_MAPPING_BOOK);
-    }
-
-    @GetMapping("/keywordSearchName")
-    public String keywordSearchByBookName(@RequestParam(value = "keywords", required = false) String text,
-            Model model) {
-        List<Book> searchResults = null;
-        try {
-            searchResults = searchservice.keywordSearch(text, "name");
-
-        } catch (Exception ex) {
-            // for now do nothing. Later we will throw an exception and display a message to
-            // the user
-        }
-        model.addAttribute(MODEL_ATTRIBUTE_BOOK, searchResults);
-        return createRedirectViewPath(REQUEST_MAPPING_BOOK);
-    }
-
-    @GetMapping("/wildcardSearchName")
-    public String wildcardSearchByBookName(@RequestParam(value = "keywords", required = false) String text,
-            Model model) {
-        List<Book> searchResults = null;
-        try {
-            searchResults = searchservice.wildcardSearch(text, "name");
-
-        } catch (Exception ex) {
-            // for now do nothing. Later we will throw an exception and display a message to
-            // the user
-        }
-        model.addAttribute(MODEL_ATTRIBUTE_BOOK, searchResults);
-        return createRedirectViewPath(REQUEST_MAPPING_BOOK);
+        if (searchResults != null && searchResults.isEmpty())
+            model.addAttribute(MODEL_ATTRIBUTE_BOOK, null);
+        else
+            model.addAttribute(MODEL_ATTRIBUTE_BOOK, searchResults);
+        return "index";
     }
 
     private String createRedirectViewPath(String requestMapping) {

--- a/src/main/java/amazin/model/Book.java
+++ b/src/main/java/amazin/model/Book.java
@@ -12,8 +12,13 @@ import org.hibernate.search.annotations.Indexed;
 
 @Entity
 @Indexed
-@Table(name="book")
+@Table(name = "book")
 public class Book {
+    public static final String BOOK_NAME_FIELD = "name";
+    public static final String BOOK_DESCRIPTION_FIELD = "description";
+    public static final String BOOK_ISBN_FIELD = "ISBN";
+    public static final String BOOK_AUTHOR_FIELD = "author";
+    public static final String BOOK_PUBLISHER_FIELD = "publisher";
 
     @Id
     @GeneratedValue(strategy = GenerationType.AUTO)
@@ -21,38 +26,38 @@ public class Book {
     private Long id;
     @NotNull
     @Size(min = 2, message = "Book name should have at least 2 characters")
-    @Column(name="book_name")
+    @Column(name = "book_name")
     @Field
     private String name;
     @NotNull
     @Size(min = 2, message = "Book description should have at least 2 characters")
-    @Column(name="book_description")
+    @Column(name = "book_description")
     @Field
     private String description;
     @NotNull
     @Size(min = 13, max = 13, message = "ISBN should be 13 characters long") // ISBN is a 13 digit number
-    @Column(name="book_isbn")
+    @Column(name = "book_isbn")
     @Field
     private String ISBN;
-    @Column(name="book_picture")
+    @Column(name = "book_picture")
     @Field
     private String picture;
     @NotNull
     @Size(min = 2, message = "Book author should have at least 2 characters")
-    @Column(name="book_author")
+    @Column(name = "book_author")
     @Field
     private String author;
     @NotNull
     @Size(min = 1, message = "Book publisher should have at least 2 characters")
-    @Column(name="book_publisher")
+    @Column(name = "book_publisher")
     @Field
     private String publisher;
     @Min(0)
-    @Column(name="book_inventory")
+    @Column(name = "book_inventory")
     @Field
     private int inventory;
     @Min(0)
-    @Column(name="book_price")
+    @Column(name = "book_price")
     @Field
     private double price;
 

--- a/src/main/java/amazin/service/HibernateSearchService.java
+++ b/src/main/java/amazin/service/HibernateSearchService.java
@@ -5,97 +5,27 @@ import org.apache.lucene.search.Query;
 import org.hibernate.search.jpa.FullTextEntityManager;
 import org.hibernate.search.jpa.Search;
 import org.hibernate.search.query.dsl.QueryBuilder;
-import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
 import javax.persistence.EntityManager;
 import javax.persistence.NoResultException;
+import javax.persistence.PersistenceContext;
+
 import java.util.List;
 
 @Service
+@Transactional
 public class HibernateSearchService {
 
-    private final EntityManager entityManager;
+    @PersistenceContext
+    private EntityManager entityManager;
 
-    @Autowired
-    public HibernateSearchService(EntityManager entityManager) {
-        super();
-        this.entityManager = entityManager;
-    }
-
-    @Transactional
-    public List<Book> fuzzySearch(String searchTerm, String fieldName) {
+    public List<Book> fuzzySearch(String searchTerm, String[] fieldNames) {
 
         FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(entityManager);
         QueryBuilder qb = fullTextEntityManager.getSearchFactory().buildQueryBuilder().forEntity(Book.class).get();
-        Query luceneQuery = qb.keyword().fuzzy().withEditDistanceUpTo(2).withPrefixLength(0).onFields(fieldName)
-                .matching(searchTerm).createQuery();
-
-        javax.persistence.Query jpaQuery = fullTextEntityManager.createFullTextQuery(luceneQuery, Book.class);
-
-        List<Book> bookList = null;
-        try {
-            bookList = jpaQuery.getResultList();
-        } catch (NoResultException nre) {
-            // for now do nothing. Later we will throw an exception and display a message to
-            // the user
-        }
-
-        return bookList;
-    }
-
-    @Transactional
-    public List<Book> keywordSearch(String searchTerm, String fieldName) {
-
-        FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(entityManager);
-        QueryBuilder qb = fullTextEntityManager.getSearchFactory().buildQueryBuilder().forEntity(Book.class).get();
-        Query luceneQuery = qb.keyword().onField(fieldName).matching(searchTerm).createQuery();
-
-        javax.persistence.Query jpaQuery = fullTextEntityManager.createFullTextQuery(luceneQuery, Book.class);
-
-        List<Book> bookList = null;
-        try {
-            bookList = jpaQuery.getResultList();
-        } catch (NoResultException nre) {
-            // for now do nothing. Later we will throw an exception and display a message to
-            // the user
-
-        }
-
-        return bookList;
-    }
-
-    @Transactional
-    public List<Book> wildcardSearch(String searchTerm, String fieldName) {
-
-        FullTextEntityManager fullTextEntityManager = Search.getFullTextEntityManager(entityManager);
-        QueryBuilder qb = fullTextEntityManager.getSearchFactory().buildQueryBuilder().forEntity(Book.class).get();
-
-        StringBuilder cleanedSearchTerm = new StringBuilder(searchTerm);
-
-        // Removs leading and trailing special characters (?, *) to improve search
-        // efficiency
-        if (cleanedSearchTerm.charAt(0) == '?' || cleanedSearchTerm.charAt(0) == '*')
-            cleanedSearchTerm.deleteCharAt(0);
-        if (cleanedSearchTerm.charAt(cleanedSearchTerm.length() - 1) == '?'
-                || cleanedSearchTerm.charAt(cleanedSearchTerm.length() - 1) == '*')
-            cleanedSearchTerm.deleteCharAt(cleanedSearchTerm.length() - 1);
-
-        // Replaces all * with ? because a * represents any character sequence while ?
-        // represents a single character in the search function so if the search term
-        // happened to have a * we want to replace it with ? to compensate for that one
-        // * without affecting the rest string
-        for (int i = 0; i < cleanedSearchTerm.length(); i++) {
-            if (cleanedSearchTerm.charAt(i) == '*')
-                cleanedSearchTerm.setCharAt(i, '?');
-        }
-
-        // Adds the * at the end of the string to match any string that starts with the
-        // search term provided
-        cleanedSearchTerm.append('*');
-
-        Query luceneQuery = qb.keyword().wildcard().onField(fieldName).matching(cleanedSearchTerm.toString())
+        Query luceneQuery = qb.keyword().fuzzy().withEditDistanceUpTo(2).onFields(fieldNames).matching(searchTerm)
                 .createQuery();
 
         javax.persistence.Query jpaQuery = fullTextEntityManager.createFullTextQuery(luceneQuery, Book.class);

--- a/src/main/resources/public/js/book_view.js
+++ b/src/main/resources/public/js/book_view.js
@@ -1,3 +1,7 @@
-$(document).ready(function(){
-        $('#book_table').DataTable();
+$(document).ready(function () {
+        $('#book_table').DataTable({
+                "language": {
+                        "search": "Filter"
+                }
+        });
 });

--- a/src/main/resources/templates/create-book.html
+++ b/src/main/resources/templates/create-book.html
@@ -5,7 +5,7 @@
 <body>
     <div layout:fragment="content">
         <h1>Create Book</h1>
-        <form action="#" th:action="@{/addbook}" th:object="${book}" method="post">
+        <form action="#" th:action="@{/addbook}" th:object="${books}" method="post">
             <p>
                 <label for="name">Name:</label>
                 <input id="name" type="text" minlength="2" th:field="*{name}">

--- a/src/main/resources/templates/layout.html
+++ b/src/main/resources/templates/layout.html
@@ -11,14 +11,14 @@
 
     <title>Amazin</title>
     <!-- Custom fonts for this template-->
-    <link rel="stylesheet" type="text/css"
-          href="/webjars/font-awesome/5.7.2/css/all.min.css" />
+    <link rel="stylesheet" type="text/css" href="/webjars/font-awesome/5.7.2/css/all.min.css" />
 
-    <link href="https://fonts.googleapis.com/css?family=Nunito:200,200i,300,300i,400,400i,600,600i,700,700i,800,800i,900,900i" rel="stylesheet">
+    <link
+        href="https://fonts.googleapis.com/css?family=Nunito:200,200i,300,300i,400,400i,600,600i,700,700i,800,800i,900,900i"
+        rel="stylesheet">
 
     <!-- Custom styles for this template-->
-    <link rel="stylesheet"
-          href="/webjars/startbootstrap-sb-admin-2/4.0.2/css/sb-admin-2.min.css" />
+    <link rel="stylesheet" href="/webjars/startbootstrap-sb-admin-2/4.0.2/css/sb-admin-2.min.css" />
 
     <link href="https://cdn.datatables.net/1.10.19/css/jquery.dataTables.min.css" rel="stylesheet" type="text/css" />
 
@@ -28,157 +28,169 @@
 
 <body id="page-top">
 
-<!-- Page Wrapper -->
-<div id="wrapper" class="h-100">
-    <!-- Content Wrapper -->
-    <div id="content-wrapper">
+    <!-- Page Wrapper -->
+    <div id="wrapper" class="h-100">
+        <!-- Content Wrapper -->
+        <div id="content-wrapper">
 
-        <!-- Main Content -->
-        <div id="content">
+            <!-- Main Content -->
+            <div id="content">
 
-            <!-- Topbar -->
-            <nav class="navbar navbar-expand navbar-light bg-white topbar mb-4 static-top shadow">
+                <!-- Topbar -->
+                <nav class="navbar navbar-expand navbar-light bg-white topbar mb-4 static-top shadow">
 
-                <!-- Icon  -->
-                <li id="icon" class="navbar-nav nav-item">
-                    <a href="/"><img id= "" src="/images/icon.png"></a>
-                </li>
-
-                <div class="topbar-divider d-none d-sm-block"></div>
-
-                <!-- Topbar Search -->
-                <form class="d-none d-sm-inline-block form-inline navbar-search">
-                    <div class="input-group">
-                        <input type="text" class="form-control bg-light border-0 small" placeholder="Search for..." aria-label="Search" aria-describedby="basic-addon2">
-                        <div class="input-group-append">
-                            <button class="btn btn-primary" type="button">
-                                <i class="fas fa-search fa-sm"></i>
-                            </button>
-                        </div>
-                    </div>
-                </form>
-
-                <!-- Topbar Navbar -->
-                <ul class="navbar-nav ml-auto">
-                    <!-- Nav Item - Search Dropdown (Visible Only XS) -->
-                    <li class="nav-item dropdown no-arrow d-sm-none">
-                        <a class="nav-link dropdown-toggle" href="#" id="searchDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
-                            <i class="fas fa-search fa-fw"></i>
-                        </a>
-                        <!-- Dropdown - Messages -->
-                        <div class="dropdown-menu dropdown-menu-right p-3 shadow animated--grow-in" aria-labelledby="searchDropdown">
-                            <form class="form-inline mr-auto w-100 navbar-search">
-                                <div class="input-group">
-                                    <input type="text" class="form-control bg-light border-0 small" placeholder="Search for..." aria-label="Search" aria-describedby="basic-addon2">
-                                    <div class="input-group-append">
-                                        <button class="btn btn-primary" type="button">
-                                            <i class="fas fa-search fa-sm"></i>
-                                        </button>
-                                    </div>
-                                </div>
-                            </form>
-                        </div>
-                    </li>
-
-                    <!-- Nav Item - Alerts -->
-                    <li class="nav-item mx-1">
-                        <a class="nav-link" href="/addbook" role="button" data-toggle="tooltip" title="Create a new book">
-                            <i class="fas fa-plus"></i>
-                        </a>
-                    </li>
-
-                    <!-- Nav Item - Messages -->
-                    <li class="nav-item dropdown no-arrow mx-1">
-                        <a class="nav-link dropdown-toggle" href="#" id="messagesDropdown" role="button" data-toggle="tooltip" title="View shopping cart">
-                            <i class="fas fa-shopping-cart"></i>
-                            <!-- Counter - Messages -->
-                            <!--<span class="badge badge-danger badge-counter">7</span>-->
-                        </a>
+                    <!-- Icon  -->
+                    <li id="icon" class="navbar-nav nav-item">
+                        <a href="/"><img id="" src="/images/icon.png"></a>
                     </li>
 
                     <div class="topbar-divider d-none d-sm-block"></div>
 
-                    <!-- Nav Item - User Information -->
-                    <li class="nav-item dropdown no-arrow">
-                        <a class="nav-link dropdown-toggle" href="#" id="userDropdown" role="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
-                            <span class="mr-2 d-none d-lg-inline text-gray-600 small">Valerie Luna</span>
-                            <img class="img-profile rounded-circle" src="https://source.unsplash.com/QAB-WJcbgJk/60x60">
-                        </a>
-                        <!-- Dropdown - User Information -->
-                        <div class="dropdown-menu dropdown-menu-right shadow animated--grow-in" aria-labelledby="userDropdown">
-                            <a class="dropdown-item" href="#">
-                                <i class="fas fa-user fa-sm fa-fw mr-2 text-gray-400"></i>
-                                Profile
-                            </a>
-                            <a class="dropdown-item" href="#">
-                                <i class="fas fa-list fa-sm fa-fw mr-2 text-gray-400"></i>
-                                Recommendations
-                            </a>
-                            <div class="dropdown-divider"></div>
-                            <a class="dropdown-item" href="#" data-toggle="modal" data-target="#logoutModal">
-                                <i class="fas fa-sign-out-alt fa-sm fa-fw mr-2 text-gray-400"></i>
-                                Logout
-                            </a>
+                    <!-- Topbar Search -->
+                    <form class="d-none d-sm-inline-block form-inline navbar-search" action="/search" method="get"
+                        autocomplete="off">
+                        <div class="input-group">
+                            <input name="keywords" type="text" class="form-control bg-light border-0 small" placeholder="Search for..."
+                                aria-label="Search" aria-describedby="basic-addon2">
+                            <div class="input-group-append">
+                                <button class="btn btn-primary" type="submit">
+                                    <i class="fas fa-search fa-sm"></i>
+                                </button>
+                            </div>
                         </div>
-                    </li>
+                    </form>
 
-                </ul>
+                    <!-- Topbar Navbar -->
+                    <ul class="navbar-nav ml-auto">
+                        <!-- Nav Item - Search Dropdown (Visible Only XS) -->
+                        <li class="nav-item dropdown no-arrow d-sm-none">
+                            <a class="nav-link dropdown-toggle" href="#" id="searchDropdown" role="button"
+                                data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                                <i class="fas fa-search fa-fw"></i>
+                            </a>
+                            <!-- Dropdown - Messages -->
+                            <div class="dropdown-menu dropdown-menu-right p-3 shadow animated--grow-in"
+                                aria-labelledby="searchDropdown">
+                                <form class="form-inline mr-auto w-100 navbar-search">
+                                    <div class="input-group">
+                                        <input type="text" class="form-control bg-light border-0 small"
+                                            placeholder="Search for..." aria-label="Search"
+                                            aria-describedby="basic-addon2">
+                                        <div class="input-group-append">
+                                            <button class="btn btn-primary" type="button">
+                                                <i class="fas fa-search fa-sm"></i>
+                                            </button>
+                                        </div>
+                                    </div>
+                                </form>
+                            </div>
+                        </li>
 
-            </nav>
-            <!-- End of Topbar -->
+                        <!-- Nav Item - Alerts -->
+                        <li class="nav-item mx-1">
+                            <a class="nav-link" href="/addbook" role="button" data-toggle="tooltip"
+                                title="Create a new book">
+                                <i class="fas fa-plus"></i>
+                            </a>
+                        </li>
 
-            <!-- Begin Page Content -->
-            <div class="container h-100">
-                <div layout:fragment="content"></div>
+                        <!-- Nav Item - Messages -->
+                        <li class="nav-item dropdown no-arrow mx-1">
+                            <a class="nav-link dropdown-toggle" href="#" id="messagesDropdown" role="button"
+                                data-toggle="tooltip" title="View shopping cart">
+                                <i class="fas fa-shopping-cart"></i>
+                                <!-- Counter - Messages -->
+                                <!--<span class="badge badge-danger badge-counter">7</span>-->
+                            </a>
+                        </li>
+
+                        <div class="topbar-divider d-none d-sm-block"></div>
+
+                        <!-- Nav Item - User Information -->
+                        <li class="nav-item dropdown no-arrow">
+                            <a class="nav-link dropdown-toggle" href="#" id="userDropdown" role="button"
+                                data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                                <span class="mr-2 d-none d-lg-inline text-gray-600 small">Valerie Luna</span>
+                                <img class="img-profile rounded-circle"
+                                    src="https://source.unsplash.com/QAB-WJcbgJk/60x60">
+                            </a>
+                            <!-- Dropdown - User Information -->
+                            <div class="dropdown-menu dropdown-menu-right shadow animated--grow-in"
+                                aria-labelledby="userDropdown">
+                                <a class="dropdown-item" href="#">
+                                    <i class="fas fa-user fa-sm fa-fw mr-2 text-gray-400"></i>
+                                    Profile
+                                </a>
+                                <a class="dropdown-item" href="#">
+                                    <i class="fas fa-list fa-sm fa-fw mr-2 text-gray-400"></i>
+                                    Recommendations
+                                </a>
+                                <div class="dropdown-divider"></div>
+                                <a class="dropdown-item" href="#" data-toggle="modal" data-target="#logoutModal">
+                                    <i class="fas fa-sign-out-alt fa-sm fa-fw mr-2 text-gray-400"></i>
+                                    Logout
+                                </a>
+                            </div>
+                        </li>
+
+                    </ul>
+
+                </nav>
+                <!-- End of Topbar -->
+
+                <!-- Begin Page Content -->
+                <div class="container h-100">
+                    <div layout:fragment="content"></div>
+                </div>
+                <!-- /.container-fluid -->
+
             </div>
-            <!-- /.container-fluid -->
+            <!-- End of Main Content -->
 
         </div>
-        <!-- End of Main Content -->
+        <!-- End of Content Wrapper -->
 
     </div>
-    <!-- End of Content Wrapper -->
+    <!-- End of Page Wrapper -->
 
-</div>
-<!-- End of Page Wrapper -->
+    <!-- Scroll to Top Button-->
+    <a class="scroll-to-top rounded" href="#page-top">
+        <i class="fas fa-angle-up"></i>
+    </a>
 
-<!-- Scroll to Top Button-->
-<a class="scroll-to-top rounded" href="#page-top">
-    <i class="fas fa-angle-up"></i>
-</a>
-
-<!-- Logout Modal-->
-<div class="modal fade" id="logoutModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
-    <div class="modal-dialog" role="document">
-        <div class="modal-content">
-            <div class="modal-header">
-                <h5 class="modal-title" id="exampleModalLabel">Ready to Leave?</h5>
-                <button class="close" type="button" data-dismiss="modal" aria-label="Close">
-                    <span aria-hidden="true">×</span>
-                </button>
-            </div>
-            <div class="modal-body">Select "Logout" below if you are ready to end your current session.</div>
-            <div class="modal-footer">
-                <button class="btn btn-secondary" type="button" data-dismiss="modal">Cancel</button>
-                <a class="btn btn-primary" href="login">Logout</a>
+    <!-- Logout Modal-->
+    <div class="modal fade" id="logoutModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel"
+        aria-hidden="true">
+        <div class="modal-dialog" role="document">
+            <div class="modal-content">
+                <div class="modal-header">
+                    <h5 class="modal-title" id="exampleModalLabel">Ready to Leave?</h5>
+                    <button class="close" type="button" data-dismiss="modal" aria-label="Close">
+                        <span aria-hidden="true">×</span>
+                    </button>
+                </div>
+                <div class="modal-body">Select "Logout" below if you are ready to end your current session.</div>
+                <div class="modal-footer">
+                    <button class="btn btn-secondary" type="button" data-dismiss="modal">Cancel</button>
+                    <a class="btn btn-primary" href="login">Logout</a>
+                </div>
             </div>
         </div>
     </div>
-</div>
 
-<!-- Bootstrap core JavaScript-->
-<script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/jquery/jquery.min.js"></script>
-<script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/bootstrap/js/bootstrap.bundle.min.js"></script>
+    <!-- Bootstrap core JavaScript-->
+    <script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/jquery/jquery.min.js"></script>
+    <script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/bootstrap/js/bootstrap.bundle.min.js"></script>
 
-<!-- Core plugin JavaScript-->
-<script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/jquery-easing/jquery.easing.min.js"></script>
+    <!-- Core plugin JavaScript-->
+    <script src="/webjars/startbootstrap-sb-admin-2/4.0.2/vendor/jquery-easing/jquery.easing.min.js"></script>
 
-<!-- Custom scripts for all pages-->
-<script src="/webjars/startbootstrap-sb-admin-2/4.0.2/js/sb-admin-2.min.js"></script>
+    <!-- Custom scripts for all pages-->
+    <script src="/webjars/startbootstrap-sb-admin-2/4.0.2/js/sb-admin-2.min.js"></script>
 
-<script src="https://cdn.datatables.net/1.10.19/js/jquery.dataTables.min.js"></script>
-<script src="/js/layout.js"></script>
-<th:block layout:fragment="script"></th:block>
+    <script src="https://cdn.datatables.net/1.10.19/js/jquery.dataTables.min.js"></script>
+    <script src="/js/layout.js"></script>
+    <th:block layout:fragment="script"></th:block>
 </body>
 
 </html>

--- a/src/main/resources/templates/update-book.html
+++ b/src/main/resources/templates/update-book.html
@@ -5,7 +5,7 @@
 <body>
 <div layout:fragment="content">
         <h1>Edit Book</h1>
-        <form action="#" th:action="@{/addbook}" th:object="${book}" method="post">
+        <form action="#" th:action="@{/update/{id}(id=${books.id})}" th:object="${books}" method="post">
             <p>
                 <label for="name">Name:</label>
                 <input id="name" type="text" minlength="2" th:field="*{name}">

